apps/fxc-front/src/app/logic/score/scorer.ts

```diff
@@ -1,19 +1,105 @@
-import { CircuitType } from '@flyxc/optimizer/lib/api';
-
-export class Score {
-  distanceM: number;
-  indexes: number[];
-  multiplier: number;
-  circuit: CircuitType;
-  closingRadiusKm: number;
-  points: number;
-
-  constructor(score: Partial<Score>) {
-    this.distanceM = score.distanceM ?? 0;
-    this.indexes = score.indexes ?? [];
-    this.multiplier = score.multiplier ?? 1;
-    this.circuit = score.circuit ?? CircuitType.OpenDistance;
-    this.closingRadiusKm = score.closingRadiusKm ?? 0;
-    this.points = score.points ?? 0;
+import type { LatLonAltTime, ScoringResult } from '@flyxc/optimizer/lib/optimizer';
+
+import { type Request as WorkerRequest, type Response as WorkerResponse } from '../../workers/optimizer';
+import ScoringWorker from '../../workers/optimizer?worker';
+import { getScoringRuleName, type LeagueCode } from './league/leagues';
+
+export enum ScoreOrigin {
+  INTERACTIVE = 'interactive',
+  TRACK = 'track',
+}
+
+export type Score = ScoringResult & { origin: ScoreOrigin };
+
+export type ScoringResultHandler = (result: ScoringResult) => void;
+
+export type ScoringRequestIdProvider = () => number;
+
+export class Scorer {
+  private scoringWorker?: Worker;
+  private readonly handleScoringResult: ScoringResultHandler;
+  private readonly getScoringRequestId: ScoringRequestIdProvider;
+
+  /**
+   * <pre>handleScoringResult</pre> and <pre>getScoringRequestId</pre> functions are invoked in a
+   * worker context which means that 'this' keyword is a reference to the worker itself.
+   * If a function body uses 'this' keyword  and is sent to the constructor as a reference, it will not work.
+   * In this case, the function should be wrapped (see example bellow).<br/>
+   *
+   * E.g:
+   * <pre>
+   * class MyClass {
+   * ...
+   *   handleResult(result: ScoringResult){
+   *   ...
+   *     this.doSomethingWithResult(result);
+   *   ...
+   *   }
+   *   ...
+   *   // although this is a valid code, it will not work because 'this' in handleResult represents the worker and then
+   *   // 'this.doSomethingWithResult(result)' will fail with this.doSomethingWithResult is not a function.
+   *   scorer = new Scorer(this.handleResult,...);
+   *   // the correct syntax is:
+   *   scorer = new Scorer((result)=>this.handleResult(result),...);
+   * }
+   * </pre>
+   * @param handleScoringResult {ScoringResultHandler}
+   *        Takes the ScoringResult into account
+   * @param getScoringRequestId {ScoringRequestIdProvider}
+   *        Get a scoring request identifier. The ScoringResultHandler is invoked only if the returned
+   *        request identifier matchs the request identifier returned by the underlying optimizer.
+   */
+  constructor(handleScoringResult: ScoringResultHandler, getScoringRequestId: ScoringRequestIdProvider) {
```

Comment:
this could be written as 

```
constructor(private handleScoringResult: ScoringResultHandler, private getScoringRequestId: ScoringRequestIdProvider) {}`
```

and you can drop the member variable declrations.